packages/api/src/datasources/reading_progress_data_source.ts

```diff
@@ -0,0 +1,84 @@
+import { redisDataSource } from '../redis_data_source'
+
+type ReadingProgressCacheItem = {
+  readingProgressPercent: number
+  readingProgressTopPercent: number | undefined
+  readingProgressAnchorIndex: number | undefined
+  updatedAt: Date
+}
+
+export class ReadingProgressDataSource {
+  private cacheItems: { [id: string]: ReadingProgressCacheItem } = {}
+
+  async getReadingProgress(
+    libraryItemID: string
+  ): Promise<ReadingProgressCacheItem | undefined> {
+    const cacheKey = `omnivore:reading-progress:${libraryItemID}`
+    const cached = this.cacheItems[cacheKey]
+    if (cached) {
+      return cached
+    }
+    return this.valueFromRedis(libraryItemID)
+  }
+
+  async updateReadingProgress(
+    libraryItemID: string,
+    progress: {
+      readingProgressPercent: number
+      readingProgressTopPercent: number | undefined | null
+      readingProgressAnchorIndex: number | undefined | null
+    }
+  ): Promise<void> {
+    const cacheKey = `omnivore:reading-progress:${libraryItemID}`
+    const existingItem = await this.valueFromRedis(cacheKey)
+    const cacheItem = {
+      readingProgressPercent: Math.max(
+        progress.readingProgressPercent,
+        existingItem?.readingProgressPercent ?? 0
+      ),
+      readingProgressTopPercent: Math.max(
+        progress.readingProgressTopPercent ?? 0,
+        existingItem?.readingProgressTopPercent ?? 0
+      ),
+      readingProgressAnchorIndex: Math.max(
+        progress.readingProgressAnchorIndex ?? 0,
+        existingItem?.readingProgressAnchorIndex ?? 0
+      ),
+      updatedAt: new Date(),
+    }
+
+    this.cacheItems[cacheKey] = cacheItem
+    if (await redisDataSource.redisClient?.hmset(cacheKey, cacheItem)) {
+      console.log('cached reading progress')
+    } else {
+      console.log('failed to cache reading progress')
+    }
+  }
```

Comment:
@jacksonh i'm not sure if concurrency here will be a problem.

this could be rare but if two updates of reading progress happen at the same time and probably the old value could be used.

Reply:
yeah i agree, the whole thing feels a little brittle.